### ISSUE TITLE
match dataclasses module api (missing fields, make_dataclass, asdict, is_dataclass) more closely

### ISSUE THREAD

**Issue by bionicles** [pydantic/pydantic#10040]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

A lot of code I have depends upon `dataclasses` because it's the closest thing to a `struct` in Python, and numerous functions in this valuable stdlib module are missing or renamed in the `pydantic.dataclasses` module, which causes it to not live up to its intended purpose to be a **drop-in replacement for `dataclasses`** -- isn't there more to it than the `@dataclass` decorator ?

- [ ] `fields` missing
- [ ] `asdict` missing -> opportunity to fix `model_dump_json` for dataclasses ?
- [ ] `make_dataclass` missing
- [ ] `is_dataclass` was renamed to `is_pydantic_dataclass`, may we alias this ?

Right now, we can't really claim the `pydantic.dataclasses` is a "drop in replacement" for the `dataclasses` module because these functions are missing or renamed, so I propose adding them and aliasing `is_pydantic_dataclass` as `is_dataclass`. 

**The benefit would be, codebases which make more extensive use of the `dataclasses` module beyond just the decorator can more easily switch to pydantic. Pydantic users could benefit from consistent & familiar helper functions to work with these types.** 

For `make_dataclass`:

Apparently there is a workaround for `make_dataclass` but it's implicit here and not explicit I.E. even if that works, everyone has to independently go and find this closed issue and re-invent the wheel on a critical path to make a dataclass. 
https://github.com/pydantic/pydantic/issues/1954

TLDR: if one person figures out `make_dataclass`, `fields`, `asdict`, `is_dataclass` then many people can benefit. I don't think a hodge podge of workarounds and reimplementations scattered across various closed issues is sufficient to consider the `pydantic.dataclasses` to be a "True" drop in replacement for `dataclasses` (YET) 

Yeah, admittedly BaseModel is more commonly used, but dataclass is simpler and is inspired by a standard library module, that simplicity and standardization also has value and is worthy of our time and effort to raise the standard. 

![image](https://github.com/user-attachments/assets/0aea8c4a-a735-4585-b7d9-707c85808e53)

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [X] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Viicos** [comment]
`fields` and `asdict` should be working with Pydantic dataclasses (we could/should document this), so I don't see the point of adding them in the `pydantic.dataclasses` module as an alias/reexport of the stdlib ones.

> * `is_dataclass` was renamed to `is_pydantic_dataclass`, may we alias this ?

I think it makes more sense to keep it named `is_pydantic_dataclass`, this is what it does (and will return `False` for stdlib dataclasses).

`make_dataclass` could make sense, although it could be solved easily with https://github.com/python/cpython/issues/118974.

**Comment by Viicos** [comment]
I opened https://github.com/python/cpython/pull/122723 to have a way to specify the dataclass factory.

**Comment by bionicles** [comment]
Ok, makes sense, I'd still suggest the re-export because the module is marketed as a "drop in replacement" so having the aliases because could save numerous users a lot of time. 

"drop in replacement for dataclasses" implies we can use it in place of that module; 

with missing functions, we need to refactor code to use either stdlib dataclasses or pydantic dataclasses (identically named) for various functions, which means it's not a drop in replacement. 

If re-exporting avoids needing to juggle two independent dataclasses modules and remember which has what, I vote for re-exporting

**Comment by bionicles** [comment]
Maybe I'm wrong, I thought I saw that somewhere, if it's not a goal then that's fine, just good. Ah, here's where I found that, Stack Overflow !![image](https://github.com/user-attachments/assets/51509c3e-4c1c-4fe7-89a6-2c72a9d525c7)

**Comment by Viicos** [comment]
https://github.com/python/cpython/pull/122723 was merged, introducing a `decorator` argument to `make_dataclass` in Python 3.14.
